noma/bitcoind.py

```diff
@@ -212,99 +221,92 @@ def generate_rpcauth(username, password=""):
     """
-    Parse key-value config files and print out values
+    Parse ini-style config files and print out values
 
     :param key: left part of key value pair
     :param config_path: path to file
     :return: value of key
     """
-    import itertools
-    import configparser
-
-    parser = configparser.ConfigParser(strict=False)
-    with open(config_path) as lines:
-        lines = itertools.chain(
-            ("[main]",), lines
-        )  # workaround: prepend dummy section
-        parser.read_file(lines)
-        return parser.get("main", key)
+    try:
+        config = IniConfig(config_path, file_error=True, encoding='utf-8')
+    except SyntaxError as error:
+        print("⚠️ " + str(error))
+    else:
+        if section:
+            return config[section][key]
+        try:
+            value = config[key]
+        except KeyError:
+            print("❌ key '{k}' does not exist".format(k=key))
+        else:
+            return value
 
 
-def set_kv(key, value, config_path):
+def set_ini_kv(key, value, config_path, section):
     """
     Set key to value in path
     kv pairs are separated by "="
 
     :param key: key to set
     :param value: value to set
     :param config_path: config file path
-    :return str: string written
+    :param section: config section
     """
-    from fileinput import FileInput
+    try:
+        config = IniConfig(config_path, file_error=True, encoding='utf-8')
+    except SyntaxError as error:
+        print("⚠️ " + str(error))
+    else:
+        if section:
+            config[section][key] = value
+        else:
+            config[key] = value
+        config.write(delimiter='=')
 
-    path = pathlib.Path(config_path)
-    config_exists = path.is_file()
 
-    if not config_exists:
-        # create empty config file
-        path.touch()
+def get_toml_kv(config_path, key, section=""):
+    """get toml key value"""
+    config = toml.load(config_path)
+    if section:
+        value = config[section][key]
+    else:
+        value = config[key]
+    return value
 
-    current_val = None
-    try:
-        current_val = get_kv(key, config_path)
-    except Exception as err:
-        print(err)
-    if value == current_val:
-        # nothing to do
-        print("{k} already set to {v}".format(k=key, v=value))
-        return
-    if current_val is None:
-        # key does not exist yet
-        with open(config_path, "a") as file:
-            # append kv pair to file
-            file.write("\n{k}={v}".format(k=key, v=value))
+
+def set_toml_kv(config_path, key, value, section=""):
+    """set toml key value"""
+    config = toml.load(config_path)
+    if section:
+        config[section][key] = value
     else:
-        with FileInput(config_path, inplace=True, backup=".bak") as file:
-            for line in file:
-                print(line.replace(current_val, value), end="")
+        config[key] = value
+    with open(config_path, "w") as toml_file:
```

Comment:
What does `with` do?

Reply:
`with` ensures the file handler is closed automatically and otherwise acts as a "try/except" block.